### ISSUE TITLE
run example.exe on Windows 10 output error msg: select, An operation was attempted on something that is not a socket

### ISSUE THREAD

**Issue by AlexiaChen** [LibVNC/libvncserver#165]
Hi Eeveryone:

### System Enviroment:      
- My Operating System is :  windows 10 enterprise 2016 LTSB
- My build toolchain: MSVC (Visual Studio 2013), Cmake

### Promblem Description
I just simply build libvncserver on windows 10 referred to: https://github.com/LibVNC/libvncserver/blob/master/.appveyor.yml
 it build compeletely success with 0 errors.

I find example.exe in the folder of examples, and run example.exe from CMD. but it repeating output error message: **select, An operation was attempted on something that is not a socket**
![example](https://cloud.githubusercontent.com/assets/8574915/23846262/a4729f92-0807-11e7-98eb-68c63356b9d0.jpeg)
it seems like the socket handle parameter did not reference a valid socket, or for select, a member of an fd_set was not valid.

So, I just download realVNC's VNC viewer to connect localhost:5900 but it is does not work and I tried no_vnc(html5 client) in webclients folder and does not work,too.

Then, I tried simple.exe, vncev.exe looks well when no client connected to them,
but they also ouptput error message as follows when vncviewer conected:
vncev.exe:
![vncev](https://cloud.githubusercontent.com/assets/8574915/23846293/c2ca8590-0807-11e7-9d5d-b6e4618efa51.jpeg)
simple.exe:
![simple](https://cloud.githubusercontent.com/assets/8574915/23846307/d6758982-0807-11e7-8212-7cf32658ced0.jpeg)

Does anyone can solve this problem? 

Finally, I tried use SDLviewer in client_examples folder, the server(vncev,exe, simple.exe,  example.exe) also repeating show error messages like above.  

I have really no idea to solve this from now.

**Comment by bk138** [comment]
Does this happen on other Windows versions as well? I have it here, too, Windows 10.

**Comment by AlexiaChen** [comment]
@bk138 
 Thank you for your report. sorry, I just tested on windows 10. I do not know the detail on other windows version.  I will test on windows 7 tomorrow.

**Comment by AlexiaChen** [comment]
Hi @bk138 ： 
   I do the test on Windows 7 32bit without SP1(on VMware), I run vncev.exe, example.exe and simple.exe on windows 7,  and use SDLviewer on win 10 to connected to win 7 on VMware, it also repeating error message as follows, but, the error message seems CHANGED!!!!
vncev in VMware:
![vncevvm](https://cloud.githubusercontent.com/assets/8574915/23885794/acdda33a-08b0-11e7-8252-c321d7b8c1d2.jpg)
simple in VMware:
![simplevm](https://cloud.githubusercontent.com/assets/8574915/23885813/e3b6dd04-08b0-11e7-9d63-8ced3f65f6c3.jpg)
example in VMware:
![examplevm](https://cloud.githubusercontent.com/assets/8574915/23885834/0bd8d1de-08b1-11e7-8fa1-3803156dcfde.jpg)

**Comment by AlexiaChen** [comment]
Maybe, microsoft changed the error tips from windows 7 to windows 10. Because, There are same error message on windows 10 like yesterday , :)

**Comment by AlexiaChen** [comment]
@bk138   
 
I Debug the libvncserver  and Comment the code sinppet as follows:
![1](https://cloud.githubusercontent.com/assets/8574915/23930216/c8a18d74-0965-11e7-9f88-481f1cad2e9d.jpg)

then , vncev works well and running comtinue, it  does not repeating oupt error message:
**rfbProcessNewConnection: error in select**
when the new connection comes.

but a new issue pops up (windows 10 and windows 7 comes same errror):
![1](https://cloud.githubusercontent.com/assets/8574915/23930313/98a31a1a-0966-11e7-8a47-0720b4af3d48.jpg)
![2](https://cloud.githubusercontent.com/assets/8574915/23930327/a7ebb11c-0966-11e7-8ba1-49a6d89005b7.jpg)

When I use RealVNC Viewer connect the server, RTB protocal report error: bad desktop size: 25600* 25600

**Comment by AlexiaChen** [comment]
@bk138 
I fixed the bug is that [winSocket is diferrent with Linux](http://stackoverflow.com/questions/10817252/why-is-invalid-socket-defined-as-0-in-winsock2-h-c), win sock is actually usigned integer, and you assign -1 to it(wrong !  big integer, so , the error message is show not valid file descriptor) , want to represent invalid socket. but win socket's invalid socket officially offerred, its Macro INVALID_SOCKET.

**Comment by AlexiaChen** [comment]
But SDLViewer, ppmtest, TightVNCViewer  RealVNCViewer does not works as well.  
![1](https://cloud.githubusercontent.com/assets/8574915/23984778/54f123c0-0a56-11e7-83c2-49291a1825b0.jpg)

**Comment by aquawicket** [comment]
I can confirm the same issue on Win7 SP1 x64 using current libvncserver-master.
libvncserver builds with 0 errors with Cmake/VS2015. Running example code gives repeated error message with no connectivity.

> httpCheckFds: select: An operation was attempted on something that is not a socket.

Is there a known fix for this?


.

**Comment by AlexiaChen** [comment]
@aquawicket  no, there is no fix for this from now.  I tried to fix this , but its a big challenge of grasp  this whole project in limited time

**Comment by lvorobjev** [comment]
I am solved this problem under Windows XP and on MinGW GCC compiller. This problem lies in the types of data used. All of socket descriptors must be declared, as SOCKET instead int, so as in WinSock2 all of SOCKET are presented as void*, but under UNIX  they are equivalent int. When you checking expression (socket(...) >= 0) under Windows, this condition is always TRUE, because void* is an unsigned integer type. Details: https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms740516(v=vs.85).aspx .
Soon I will make a commit of my own solution of this problem.

**Comment by lvorobjev** [comment]
I present to your attention my pull request #185  and I'm waiting for your comments.
Best wishes, Lev Vorobjev.

**Comment by AlexiaChen** [comment]
@lvorobjev  Thanks , I memtioned the bug is Socket data type above,  I tried to fix it before, but the other issue pops up. so I stop research it.  Let me check your pull request.

**Comment by gops4u** [comment]
Hi,

I found another link which provides insight into a similar issue and fix, https://sourceforge.net/p/libvncserver/bugs/39/ 

@Ivorobjev : Are you able to connect to your vnc server built using mingw using the novnc?
I was able to build the vnc server using mingw64_x86_64 but web-sockets couldn't be built.

**Comment by aquawicket** [comment]
That bug is 3 years old, Has libvncserver not worked on windows for that long?  Or maybe it came incrementally with new win OS's. Anyway, What was the last version that worked on windows? I have no luck creating any server on windows.  All errors above continue.

**Comment by bk138** [comment]
I think the reason is that it was almost exlusively built with MinGW for Windows, not MSVC.

**Comment by ChongDeng** [comment]
this issue still exist. cant run on Windows 10

**Comment by aquawicket** [comment]
[whm200410](https://github.com/whm200410/libvncserver) fork works for windows 10, but has not been tested on all systems yet. I use [whm200410's](https://github.com/whm200410/libvncserver) version for windows and "this" fork for OSX and LINUX.  I think [whm200410's](https://github.com/whm200410/libvncserver) fork breaks something on unix platforms. Still needs investigating.

**Comment by aquawicket** [comment]
**Update:**   
Only LibVNC vncserver is broken.
LibVNC vncclient **works** on Windows with this code. :)
**[whm200410 fork](https://github.com/whm200410/libvncserver) fixes vncserver on Windows.**

**Comment by aquawicket** [comment]
**Finale Update:**
I can confirm that **lvorobjev** patches compile and function on Windows32, Windows64, Osx64 and Linux64 successfully. Both vncclient and vncserver are working.

**Comment by sfhacker** [comment]
We are facing the same issues here on Windows 10. This is quite an **old post**  and it (this issue) does not seem fixed. We should attempt to fix severe or critical and/or old issues first. Then move on to other issues.
Is there any patch available somewhere? We are new to this project and not much idea about how to test it (server & client) on Windows. Thanks in advance.

**Comment by bk138** [comment]
This relates to #185 which needs thorough review by someone familiar with the code and possible side effects. This will get closed once #185 or a similar PR is merged.

**Comment by Dian8** [comment]
Having "rfbProcessNewConnection: error in select: The operation completed successfully" on windows 7 too (libvncserver master)

**Comment by bk138** [comment]
Fixed by #257, thanks @tobydox
